consensus/safety-rules/src/network.rs

```diff
@@ -0,0 +1,167 @@
+        }
+        let stream = stream?;
+        stream.set_nodelay(true)?;
+
+        Ok(Self {
+            stream: NetworkStream::new(stream),
+        })
+    }
+
+    /// Blocking read until able to successfully read an entire message
+    pub fn read(&mut self) -> Result<Vec<u8>> {
+        self.stream.read()
+    }
+
+    /// Blocking read until able to successfully read an entire message
+    pub fn write(&mut self, data: &[u8]) -> Result<()> {
+        self.stream.write(data)
+    }
+}
+
+pub struct NetworkServer {
+    listener: TcpListener,
+    stream: Option<NetworkStream>,
+}
+
+impl NetworkServer {
+    pub fn new(listen: SocketAddr) -> Self {
+        let listener = TcpListener::bind(listen).unwrap();
+        Self {
+            listener,
+            stream: None,
+        }
+    }
+
+    /// If there isn't already a downstream client, it accepts. Otherwise it
+    /// blocks until able to successfully read an entire message
+    pub fn read(&mut self) -> Result<Vec<u8>> {
+        let stream = self.client()?;
+        stream.read()
+    }
+
+    pub fn write(&mut self, data: &[u8]) -> Result<()> {
+        let stream = self.client()?;
+        stream.write(data)
+    }
+
+    fn client(&mut self) -> Result<&mut NetworkStream> {
+        if self.stream.is_none() {
+            let (stream, _stream_addr) = self.listener.accept()?;
+            stream.set_nodelay(true)?;
+            self.stream = Some(NetworkStream::new(stream));
+        }
+
+        self.stream
+            .as_mut()
+            .ok_or_else(|| anyhow!("Unable to retrieve a stream"))
+    }
+}
+
+struct NetworkStream {
+    stream: TcpStream,
+    buffer: Vec<u8>,
+    temp_buffer: [u8; 1024],
+}
+
+impl NetworkStream {
+    pub fn new(stream: TcpStream) -> Self {
+        Self {
+            stream,
+            buffer: Vec::new(),
+            temp_buffer: [0; 1024],
+        }
+    }
+
+    /// Blocking read until able to successfully read an entire message
+    pub fn read(&mut self) -> Result<Vec<u8>> {
+        let result = self.read_buffer();
+        if !result.is_empty() {
+            return Ok(result);
+        }
+
+        loop {
+            let read = self.stream.read(&mut self.temp_buffer)?;
+            self.buffer.extend(self.temp_buffer[0..read].to_vec());
+            let result = self.read_buffer();
+            if !result.is_empty() {
+                return Ok(result);
+            }
+        }
+    }
+
+    /// Blocking write until able to successfully send an entire message
+    pub fn write(&mut self, data: &[u8]) -> Result<()> {
+        let u32_max = u32::max_value() as usize;
+        ensure!(
+            data.len() <= u32_max,
+            "Found data that is too large to decode: {}",
+            data.len()
+        );
+        let data_len = data.len() as u32;
+        self.write_all(&data_len.to_le_bytes())?;
+        self.write_all(data)?;
+        Ok(())
+    }
+
+    fn read_buffer(&mut self) -> Vec<u8> {
```

Comment:
Wow, I would expect there would be some libraries that would do this kind of low level logic for reading from a buffer. Is it because you don't want TCB dependencies? (what these dependencies would look like?)

Reply:
This is somewhat exploratory in context of what standard rust provides in their "pure" Networking code. we also need to add in better error handling, I should probably pepper this file with todos.

In context of other tools, we could consider mio, tokio, libp2p but each contains substantial amount of unsafe code and are rather large code bases. At least for the TCB, if we can build something that is moderately simple and reliable, we ought to go in that direction. We'll need to work with the networking team to ensure that whatever is ultimately done here is a subset of libranet.